### ISSUE TITLE
[Bug]: Casting no longer works due to media files being inacessible without auth

### ISSUE THREAD

**Issue by zmanplex** [tubearchivist/tubearchivist#331]
### Latest and Greatest

- [X] I'm running the latest version of Tube Archivist and have read the [release notes](https://github.com/tubearchivist/tubearchivist/releases/latest).

### Operating System

unraid

## Describe the bug
Casting media to a chromecast device allows a successful cast connection but media never plays

when chromecast requests the video file TA responds with a 403 error code

This is caused by the nginx config requiring a valid response from /api/ping before allowing access to /media

### Relevant log output

```shell
Oct  3 11:17:33 2022] GET 
/media/David%20Pakman%20Show/20221003_YwjoSm_U2uc_Are%20You%20Cherry-Picking%20These%20Republican%20Lunatics.mp4 
=> generated 58 bytes in 3 msecs (HTTP/1.1 403) 8 headers in 262 bytes (1 
switches on core 0)
ta_video/_doc/YwjoSm_U2uc
[pid: 52|app: 0|req: 700/700] 172.18.0.6 () {52 vars in 829 bytes} [Mon 
Oct  3 11:17:51 2022] GET /cache/videos/y/YwjoSm_U2uc.jpg => generated 58 
bytes in 3 msecs (HTTP/1.1 403) 8 headers in 262 bytes (1 switches on core 
0)
[pid: 52|app: 0|req: 701/701] 172.18.0.6 () {52 vars in 823 bytes} [Mon 
Oct  3 11:17:51 2022] GET /cache/videos/y/YwjoSm_U2uc.jpg => generated 58 
bytes in 2 msecs (HTTP/1.1 403) 8 headers in 262 bytes (1 switches on core 
0)
[pid: 52|app: 0|req: 702/702] 172.18.0.6 () {52 vars in 829 bytes} [Mon 
Oct  3 11:17:51 2022] GET /cache/videos/y/YwjoSm_U2uc.jpg => generated 58 
bytes in 2 msecs (HTTP/1.1 403) 8 headers in 262 bytes (1 switches on core 
0)
[pid: 52|app: 0|req: 703/703] 172.18.0.6 () {64 vars in 1277 bytes} [Mon 
Oct  3 11:17:52 2022] GET 
/media/David%20Pakman%20Show/20221003_YwjoSm_U2uc_Are%20You%20Cherry-Picking%20These%20Republican%20Lunatics.mp4 
=> generated 58 bytes in 5 msecs (HTTP/1.1 403) 8 headers in 262 bytes (1 
switches on core 0)
```

### Anything else?

Possible solutions: nginx workaround? python cast library? https://pypi.org/project/PyChromecast/

**Comment by bbilly1** [comment]
Excellent, thanks for collecting this here. Just to mention, without the auth request from nginx to api/ping, all user generated static files like media files and artwork, will be accessible by anyone.

That's due to the fact, that these files get served directly from nginx instead off making the redundant trip to uwsgi...

So we need some form of authentication there, to keep this project privacy friendly.

**Comment by eduncan911** [comment]
Just from an outsider's perspective...

Why not serve the binary up through the uwsgi?  Sure, not as fast as static and overhead of processing.  But you want auth, and that solves it.

Another idea is if casting can't provide the authentication you are wanting, why not move the authentication elsewhere, like to MAC addresses allowed to access those functions in nginx.  

That would easy, and very performant, to do in Lua code to check for allowed Mac addresses.  And put the configuration in a UI somewhere for that lookup table.

**Comment by bbilly1** [comment]
I appreciate your input, thanks for taking the time. What do you mean by *binary*? The artwork and media files?

I did some more researching into this one some time ago and I realized that other media server projects like Emby don't have authentication on artwork either. But their urls are build more randomly, compared to this project, where all static links are build predictably and identically on all installations.

But yeah, if we accept that if you are using cast, you won't have authentication on static files anymore, we could configure cast over an environment variable and disable the auth request in nginx.conf at startup. As long as it's clear what's happening, you can still decide if you want to put this out into the internet.

So we wouldn't have to accept the performance hit for *everybody* by serving it all through uwsgi and wouldn't have to accept the security hole for *everybody* by disabling auth.

The MAC address approach, possible but very cumbersome, as you'd have to do that for every device, some even randomize their mac address... 

Ideally, we'd also use the python cast library, to have a more native approach.

TLDR: This needs somebody with a cast device, to work on and fix, maybe you? :-) Of course I'd assist wherever I can.

**Comment by eduncan911** [comment]
I had no idea people were exposing these kind of things to the Internet.  I still do not understand the reasoning behind people exposing their home services to the internet.  The enormous risk is too great, especially from these little open source projects  that have zero exposure to pentesting, exploit reporting, and transparency.  

It's like you're taking the least supported code, with zero accountability of the developers, and no pledge to exploit testing with security bug pushes in timely manners - and putting it out on the Internet running in a userspace with permissions to access your libraries.  

Then there is the entire dependency chain pip library hacks and exploits we're dealing with at this time in cybersec (there were 68 more pypi dep libraries found this week alone that were found and disabled by security runners and cicds, some quite common packages).

/rant...  Sorry!

I explicitly use VPNs, and never expose a single port.  And I've highly advised everyone over the last two decades to never expose their services to the internet.

Even through "nginx auth" protection.

> binary

...as in assets, yes.  Typically people want to stay out of the uwsgi for assets because of "performance reasons", when in reality nginx can still serve 10s of 1000s of binary requests in the gateway.

> TLDR: This needs somebody with a cast device, to work on and fix, maybe you? :-) Of course I'd assist wherever I can.

I have a couple including the HDR 4k version.  Sure.  But I'm swapped with many other projects.  Lol

If someone wants to throw a bunch of tests out there, I'll test and report back and maybe peek at the code for tweaks. 

---

Taking a step back, if the issue is exporting assets through nginx...  Then, let the user decide on the risk or not.  

You can have two different nginx configs for your users:

- if you are exposing your nginx to the internet, then run it through the gateway for nginx auth.  It will be slower, but you'll be protected from basic discovery attacks.
- if the user wants full Chromecast support internally, and don't care about exposing assets in order to give a better user experience, then use this simple static assets nginx configs that has no auth.

Yeah, I wouldn't worry about Lua here unless you want to do custom auth at the nginx layer.

**Comment by eduncan911** [comment]
Another idea:

Have a new endpoint in nginx specifically for /chromecast or /cast that only serves up to local LAN ip mask.  That way, only media works locally - which is the intent of Chromecast anyways.

**Comment by bbilly1** [comment]
I think the above commit has fixed the issue by removing authentication on the static files. You'll have to enable cast with an env var on the Tube Archivist container: `ENABLE_CAST=True`, this will modify the nginx.conf file at startup and as far as I can tell, this will still load the JS libraries.

Can somebody verify? The fixed is pushed in the latest *unstable* build.

**Comment by null-affinity** [comment]
Pulled from unstable, and set the env var and it is now working as expected. Thank you!

**Comment by bbilly1** [comment]
@null-affinity Excellent news! Thanks a lot for testing and confirming that.

**Comment by bbilly1** [comment]
Fix is now merged in v0.3.1, please update. Thanks for everybody who helped.
